src/Npgsql/PoolManager.cs

```diff
@@ -13,20 +11,45 @@ namespace Npgsql;
 /// </remarks>
 static class PoolManager
 {
-    internal static ConcurrentDictionary<string, NpgsqlDataSource> Pools { get; } = new();
+    static ConcurrentDictionary<string, NpgsqlDataSource> _pools = new();
+
+    internal static ConcurrentDictionary<string, NpgsqlDataSource> Pools => _pools;
 
     internal static void Clear(string connString)
     {
-        // TODO: Actually remove the pools from here, #3387 (but be careful of concurrency)
-        if (Pools.TryGetValue(connString, out var pool))
-            pool.Clear();
+        var pools = _pools;
+        if (!pools.TryRemove(connString, out var pool))
+                return;
+
+        pool.Dispose();
+
+        // The same NpgsqlDataSource instance may be referenced via different connection strings
+        // in our dictionary, so we have to make sure that we remove all occurrences
+#if NET5_0_OR_GREATER
+        foreach (var pair in pools)
+            if (ReferenceEquals(pair.Value, pool))
+                pools.TryRemove(pair);
+#else
+        foreach (var pair in pools.ToArray())
+            if (ReferenceEquals(pair.Value, pool))
+                // Slightly unsafer than the modern TryRemove() overload above
+                // since it doesn't check the value for equality.
+                if (pools.TryRemove(pair.Key, out var otherPool) && !ReferenceEquals(otherPool, pool))
```

Comment:
Am not sure we need to go into the complexity of adding back to the pool etc... Since you check the pool identity before removing, the only reason TryRemove would return a different pool is in a race condition where the other pool was already removed by someone else (e.g. a concurrent racing Clear/ClearAll). That seems like a benign race condition, and I'm not sure there's any real value in re-adding it back. For simplicity, I'd just always do TryRemove and call Dispose on the result.

Also see comment below about a PoolManager write lock which would make this race condition impossible.

Reply:
Yeah, I agree. My thought was to make the observeable behavior across different .NET versions as compatible as possible but since it will never be 100% the amount of compatibility we aim at lies in the eye of the beholder and simlicity definitly is a god thing.